### PR TITLE
smt2 backend: fix floating-point values

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 84816089ad6df4578ec5e7b21be43255bad846b0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94114906

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b164d139f9bb5b939b9a116b6d1579c9cd4e9c8d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94115165

File: src/solvers/floatbv/float_bv.cpp [hunk 1]
```diff
@@ -103,21 +103,19 @@ ieee_float_spect float_bvt::get_spec(const exprt &expr)
 exprt float_bvt::abs(const exprt &op, const ieee_float_spect &spec)
 {
   // we mask away the sign bit, which is the most significant bit
-  std::string mask_str(spec.width(), '1');
-  mask_str[0]='0';
+  const mp_integer v = power(2, spec.width() - 1) - 1;
 
-  constant_exprt mask(mask_str, op.type());
+  constant_exprt mask(integer2bvrep(v, spec.width()), op.type());
```
Comment (romainbrenguier) [inline src/solvers/floatbv/float_bv.cpp]: const

**Review by romainbrenguier — APPROVED** [review]

File: src/solvers/floatbv/float_bv.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/floatbv/float_bv.cpp]: done

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5e8996706d578fe8b330013386120a465e16f155).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94360271

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3587d9d325e26dd2fc5a6a01bb8af9dc419c7b66).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94374551

File: src/solvers/floatbv/float_bv.cpp [hunk 2]
```diff
@@ -103,21 +103,19 @@ ieee_float_spect float_bvt::get_spec(const exprt &expr)
...
 
   return bitand_exprt(op, mask);
 }
 
 exprt float_bvt::negation(const exprt &op, const ieee_float_spect &spec)
 {
   // we flip the sign bit with an xor
-  std::string mask_str(spec.width(), '0');
-  mask_str[0]='1';
+  const mp_integer v = power(2, spec.width() - 1);
```
Comment (martin-cs) [inline src/solvers/floatbv/float_bv.cpp:116]: Does this mean that we can generate two NaN values that are not SMT-LIB equal?

**Review by martin-cs — APPROVED** [review]
Looks like progress to me.

File: src/solvers/floatbv/float_bv.cpp (hunk 2 above)
Comment (kroening) [reply src/solvers/floatbv/float_bv.cpp:116]: Uh -- important point. My impression was that the various (fp t #b1..1 s) would not be equal w.r.t. =?

**Pull request merged**